Auth0/SafariAuthenticationSession.swift

```diff
@@ -29,14 +29,46 @@ import AuthenticationServices
 #if swift(>=3.2)
 @available(iOS 11.0, *)
 class SafariAuthenticationSession: AuthSession {
+    private enum AuthenticationSession {
```

Comment:
Can we use a regular class instead of an enum?

Reply:
What kind of design do you have in mind using a class or struct?
Also is there a rationale for avoiding enumerations? They're a first-class type in Swift and this use-case seems like a perfect fit for an enumeration.

Reply:
I dont see it as a natural use of an enum or us having more options than these two, maybe something like this:

```swift
struct NativeAuthSession {

  @available(iOS 12.0, *)
  private var authServiceSession: ASWebAuthenticationSession?
  private var safariSession: SFAuthenticationSession?
  
  @available(iOS 12.0, *)
  init(_ session: ASWebAuthenticationSession) {
    this.authServiceSession = session;
  }
 
  init(_ session: SFAuthenticationSession) {
    this.safariSession = session;
  }

  func start() -> Bool {
    if #available(iOS 12.0, *) {
      this.authServiceSession.start();
    } else {
      this.safariServices.start();
    }
  }

  func cancel() {
    if #available(iOS 12.0, *) {
      this.authServiceSession?.cancel();
    } else {
      this.safariServices?.cancel();
    }
  }
}
```

(might have typos since I coded it on the GH UI)

But probably is going into too mich details

Reply:
The way I see it, we have the situation where we know that an object is one of two types, and that we will only have one, not both of those types of object. To me this seems like a perfect fit for an enum, since we can encode the fact that there's exactly one of those objects, and the type of that object into the type system.
The design where a struct has multiple optional properties feels like a less good fit because the type system admits the possibility that both of those properties may be nil at the same time, and requires unwrapping the optional in order to operate on the contained object.

Is there a style guide or a reason that you would prefer not to use an enum for this use case?